fdbserver/Ratekeeper.actor.cpp

```diff
@@ -840,14 +846,72 @@ ACTOR Future<Void> monitorThrottlingChanges(RatekeeperData *self) {
 	}
 }
 
+Future<Void> refreshStorageServerCommitCost(RatekeeperData *self) {
+	double elapsed = now() - self->lastBusiestCommitTagPick;
+	if(elapsed <= 0) return Void();
+	// for each SS, select the busiest commit tag from ssTagCommitCost
+	for(auto it = self->storageQueueInfo.begin(); it != self->storageQueueInfo.end(); ++it) {
+		it->value.busiestWriteTag.reset();
+		TransactionTag busiestTag;
+		TransactionCommitCostEstimation maxCost;
+		double maxRate = 0, maxBusyness = 0;
+		for(const auto& [tag, cost] : it->value.tagCostEst) {
+			// opsBeforeSample / opsSampled = COMMIT_SAMPLE_BYTE / 1
+			double rate = CLIENT_KNOBS->COMMIT_SAMPLE_BYTE * cost.getOpsSum() / elapsed;
+			if(rate > maxRate) {
+				busiestTag = tag;
+				maxRate = rate;
+				maxCost = cost;
+			}
+		}
+		if(maxRate > SERVER_KNOBS->MIN_TAG_PAGES_RATE) {
+			it->value.busiestWriteTag = busiestTag;
+			// TraceEvent("RefreshSSCommitCost").detail("TotalWriteBytes", it->value.totalWriteBytes).detail("TotalWriteOps",it->value.totalWriteOps);
+			ASSERT(it->value.totalWriteBytes > 0 && it->value.totalWriteOps > 0);
+			maxBusyness = (SERVER_KNOBS->OPERATION_COST_BYTE_FACTOR * maxCost.getOpsSum() + maxCost.getBytesSum()) /
+			              (SERVER_KNOBS->OPERATION_COST_BYTE_FACTOR * it->value.totalWriteOps + it->value.totalWriteBytes);
+			it->value.busiestWriteTagFractionalBusyness = maxBusyness;
+			it->value.busiestWriteTagRate = maxRate;
+		}
+		// reset statistics
+		it->value.tagCostEst.clear();
+		it->value.totalWriteOps = 0;
+		it->value.totalWriteBytes = 0;
+
+		TraceEvent("BusiestWriteTag", it->key)
+			.detail("Elapsed", elapsed)
+			.detail("Tag", printable(busiestTag))
+			.detail("TagOps", maxCost.getOpsSum())
+			.detail("TagBytes", maxCost.getBytesSum())
+		    .detail("TagRate", maxRate)
+		    .detail("TagBusyness", maxBusyness)
+			.detail("Reported", it->value.busiestWriteTag.present())
+			.trackLatest(it->key.toString() + "_BusiestReadTag");
+	}
+	self->lastBusiestCommitTagPick = now();
+	return Void();
+}
+
 void tryAutoThrottleTag(RatekeeperData *self, StorageQueueInfo const& ss) {
-	if(ss.busiestTag.present() && ss.busiestTagFractionalBusyness > SERVER_KNOBS->AUTO_THROTTLE_TARGET_TAG_BUSYNESS && ss.busiestTagRate > SERVER_KNOBS->MIN_TAG_COST) {
+	double busyness = 0, rate = 0;
+	Optional<TransactionTag> busiestTag;
+	if(ss.busiestWriteTag.present()) {
+		busiestTag = ss.busiestWriteTag.get();
+		busyness = ss.busiestWriteTagFractionalBusyness;
+		rate = ss.busiestWriteTagRate;
+	}
+	if(ss.busiestReadTag.present() && ss.busiestReadTagRate > rate) {
+		busiestTag = ss.busiestReadTag.get();
+		busyness = ss.busiestReadTagFractionalBusyness;
+		rate = ss.busiestReadTagRate;
+	}
```

Comment:
Are the rates being computed for reads and writes actually comparable? I wouldn't expect them to be, and if they aren't then I'm not sure we can do this comparison.

Reply:
The rate calculation of rate of read tags is
```
cost = bytes_read/16K + 1
rate = cost / sample_prob / elapsed
```
Here I understand the cost like "how many pages the sampling read involved", and rate like "how many page ops a read tag do within 1 sec". As for write, I also sample ops based on 16K (COMMIT_SAMPLE_BYTE), I think here writeRate also means "how many page ops the write tag does within 1 sec". So I suppose we can compare them.